### DISCUSSION TITLE
[Pipfile.lock] Failed to create virtual environment

### DISCUSSION THREAD

**Discussion by chauvenne** [renovatebot/renovate#10984] [Request Help]
**How are you running Renovate?**

- [ ] WhiteSource Renovate hosted app on github.com
- [x] Self hosted

Renovate version: 25.56.8

**Describe the bug**

`pipenv lock` command fails to create the virtual environment. `pipenv` raises a `ValueError: too many values to unpack (expected 2)` when calling `sys.exit(cli())`. The Python version inside the Renovate docker is 3.9. 

Our Pipfile looks like the following:
```
[[source]]
url = "https://pypi.python.org/simple"
verify_ssl = true
name = "pypi"

[packages]
black = "==21.7b0"

[requires]
python_version = "3.8"

[pipenv]
allow_prereleases = true
```

When setting `python_version = "3.9"` in the Pipfile, everything works as expected. 

**Relevant debug logs**

<details><summary>Click me to see logs</summary>

```
Command failed: pipenv lock
Creating a virtualenv for this project...
Pipfile: /builds/generic-tools/infra/renovate-runner/renovate/repos/gitlab/workers/common/vhr_imagery_tiler/Pipfile
Using /usr/bin/python3.8 (3.8.10) to create virtualenv...
⠋ Creating virtual environment...⠙ Creating virtual environment...ValueError: too many values to unpack (expected 2)

✘ Failed creating virtual environment 

[pipenv.exceptions.VirtualenvCreationException]: 
Failed to create virtual environment.
```

</details>

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [x] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

...

**Comment by rarkins** [comment]
The Renovate full image ships with latest languages/package managers, meaning python 3.9: https://github.com/renovatebot/docker-renovate-full/blob/dfc78f3c32cf5140a4dc327ab0d4d130c68887e9/Dockerfile#L67

If you need an image with an older version of Python then you need to build it yourself.

Alternatively if you have ideas for how we could build the image differently to allow dynamic python version selection then the relevant file is here: https://github.com/containerbase/buildpack/blob/main/src/usr/local/buildpack/tools/python.sh

**Reply by thib12** [reply]
The following feature from pipenv might be useful: https://pipenv-fork.readthedocs.io/en/latest/advanced.html#automatic-python-installation

**Reply by rarkins** [reply]
Maybe, if it has a non-interactive mode

**Reply by thib12** [reply]
With `PIPENV_YES` (https://github.com/pypa/pipenv/blob/master/pipenv/environments.py#L277)

**Reply by viceice** [reply]
Last time i looked at this, it would compile python on the fly, which would of cause timeout the renovate run.

@thib12 see https://github.com/containerbase/buildpack, there are issues to plan dynamic runtime install of tools. Renovate is based of that images / tools

**Reply by thib12** [reply]
It only seems weird that python3.8 is available in your image but seems to be corrupted (no pip...). This might be the root cause of the issue.

Indeed, pipenv completes successfully for a Pipfile with python_version = "3.7" if pyenv is installed in your image.

**Reply by chauvenne** [reply]
I confirm that removing `/usr/bin/python3.8` and installing `pyenv` solves the isssue.

**Reply by viceice** [reply]
The problem with the full image is, that some tools have a transitive dependency to system python, to for ubuntu focal cuurently a minimal python 3.8 is installed.

For self-hosted renovate i would always prefer to use the slim renovate with docker sidecar. This works for kubernetes too.

https://github.com/renovatebot/helm-charts/tree/main/charts/renovate#docker-in-docker-configuration

**Reply by rarkins** [reply]
Can we add pyenv as a fallback?

**Reply by viceice** [reply]
Sure, but i don't like. As i said before, it would download python source and do a compile for python versions 😕 

I think a better solution is to find the cause, why python 3.8 isn't working as expected.

We already had discussions about this multiple times and decided to use our own binaries. So we need to make the buildpack issues done, so containerbase/buildpack#101 needs to be done.

**Reply by chauvenne** [reply]
We finally switched to Docker sidecars and it works fine. Thanks for the help 😄

**Reply by viceice** [reply]
Yeah. We always prefer to use the docker mode. As that's the only way to support constraints for tool versions
